torchvision/models/detection/faster_rcnn.py

```diff
@@ -141,24 +142,41 @@ class FasterRCNN(GeneralizedRCNN):
         >>> predictions = model(x)
     """
 
-    def __init__(self, backbone, num_classes=None,
-                 # transform parameters
-                 min_size=800, max_size=1333,
-                 image_mean=None, image_std=None,
-                 # RPN parameters
-                 rpn_anchor_generator=None, rpn_head=None,
-                 rpn_pre_nms_top_n_train=2000, rpn_pre_nms_top_n_test=1000,
-                 rpn_post_nms_top_n_train=2000, rpn_post_nms_top_n_test=1000,
-                 rpn_nms_thresh=0.7,
-                 rpn_fg_iou_thresh=0.7, rpn_bg_iou_thresh=0.3,
-                 rpn_batch_size_per_image=256, rpn_positive_fraction=0.5,
-                 rpn_score_thresh=0.0,
-                 # Box parameters
-                 box_roi_pool=None, box_head=None, box_predictor=None,
-                 box_score_thresh=0.05, box_nms_thresh=0.5, box_detections_per_img=100,
-                 box_fg_iou_thresh=0.5, box_bg_iou_thresh=0.5,
-                 box_batch_size_per_image=512, box_positive_fraction=0.25,
-                 bbox_reg_weights=None):
+    def __init__(
+        self,
+        backbone: nn.Module,
+        num_classes: Optional[int] = None,
+        # transform parameters
+        min_size: int = 800,
+        max_size: int = 1333,
+        image_mean: Optional[Tuple[float]] = None,
+        image_std: Optional[Tuple[float]] = None,
+        # RPN parameters
+        rpn_anchor_generator: Optional[AnchorGenerator] = None,
+        rpn_head: Optional[nn.Module] = None,
+        rpn_pre_nms_top_n_train: int = 2000,
+        rpn_pre_nms_top_n_test: int = 1000,
+        rpn_post_nms_top_n_train: int = 2000,
+        rpn_post_nms_top_n_test: int = 1000,
+        rpn_nms_thresh: float = 0.7,
+        rpn_fg_iou_thresh: float = 0.7,
+        rpn_bg_iou_thresh: float = 0.3,
+        rpn_batch_size_per_image: int = 256,
+        rpn_positive_fraction: float = 0.5,
+        rpn_score_thresh: float = 0.0,
+        # Box parameters
+        box_roi_pool: Optional[MultiScaleRoIAlign] = None,
+        box_head: Optional[nn.Module] = None,
```

Comment:
`box_head: Optional[RPNHead] = None` I believe

Reply:
I'm unsure of what you mean @pmeier . I'm new to type hints.
I think trying to be very strict in typing by custom classes might be too much. If `nn.Module` satisfies the purpose I think it should be good enough?

Reply:
> If nn.Module satisfies the purpose I think it should be good enough?

Yes. As a rule of thumb, the input types should be as loose as possible and the output types as strict as possible. 

After some digging, I think 

https://github.com/pytorch/vision/blob/fe3118c51b487faf1aa59861ffd08b5658f41254/torchvision/models/detection/roi_heads.py#L753

is the only time we do anything with the object passed as `box_head`. Since it is possible to call an `nn.Module`, using it as annotation is fine.

---

In other cases @frgfm's comment might have some merit in other cases: for example, if we at some point access `box_head.foo`, `nn.Module` is not sufficient as annotation, since it doesn't define a `foo` attribute. In such a case we have two options:
1. Annotate with a custom class such as `RPNHead` that has this attribute.
2. Create a  `typing.Protocol` that defines the interface we are looking for. In the example above the object passed to `box_head` needs to be callable and define an attribute `foo`.

Especially if users are allowed to pass custom objects, I would always prefer 2. over 1. because it doesn't require any changes on the user side. I don't know if this works with torchscript though.

Let's discuss this if the need for something like this arises.